**<p style = "text-align: center">TRƯỜNG ĐẠI HỌC KHOA HỌC TỰ NHIÊN - ĐẠI HỌC QUỐC GIA TPHCM</p>**

# **<p style = "text-align: center">NHẬP MÔN KHOA HỌC DỮ LIỆU</p>**
## **<p style = "text-align: center">Đồ án Cuối kì</p>**
### <p style = "text-align: center"> <b>Đề tài:</b> Phân tích, khám phá dữ liệu về sách trên web thương mại điện tử (tiki)</p>
**Sinh viên thực hiện**

| Mã số sinh viên | Họ và tên |
| --- | --- |
| 20120040 | Nguyễn Quang Gia Bảo |
| 20120136 | Huỳnh Tuấn Nam |
| 20120158 | Trần Hoàng Anh Phi |

---

# **1. Thông tin về dữ liệu**<br>
Dữ liệu được sử dụng trong bài được lấy từ trang: <a href = "https://tiki.vn/sach-truyen-tieng-viet/c316"> https://tiki.vn/sach-truyen-tieng-viet/c316</a><br>

Vì website <a href = 'https://tiki.vn/'> https://tiki.vn/ </a> có khá là nhiều mặt hàng nên nhóm chỉ chọn ra danh mục <font color=red>sách tiếng Việt</font> để crawl và thực hiện đồ án này.


# **2. Khám phá dữ liệu**<br>

#### Cài đặt các thư viện cần thiết

In [1]:
# !pip install fuzzywuzzy
# !pip install python-Levenshtein

#### Import các package cần thiết

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### Đọc dữ liệu đã thu thập từ file csv 

In [3]:
df = pd.read_csv('full_data/data.csv')
print(df.shape)
display(df.head())

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_9952\906639567.py:1: DtypeWarning: Columns (22,27,29,30,31,32,33,34,35,36,37,38,39,40,41,43,45,46,47,48,49,50,51,52,53,54,56,57,58,59,60,61,62,63,64,65,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('full_data/data.csv')


(96647, 84)


Unnamed: 0         id  master_id            sku  \
0           0  204317934  204317934  9972713357587   
1           1  205279963  205279963  7618382614299   
2           2  193209826  193209826  5001141446127   
3           3  205288789  205288789  9035774965791   
4           4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  ...  kich_thuoc  tai_trong device_brand  \
0          325000             19  ...         NaN        NaN          NaN   
1          325000             21  ...         NaN        NaN          NaN   
2          129000             26  ...         NaN        NaN          NaN   
3          325000             10  ...         NaN        NaN          NaN   
4          325000             10  ...         NaN        NaN          NaN   

  do_nhay  tan_so huong_dan_bao_quan  dac_diem_noi_bat dieu_kien_su_dung  \
0     NaN     NaN                NaN               NaN               NaN   
1     NaN     NaN                NaN               NaN               NaN   
2     NaN     NaN                NaN               NaN               NaN   
3     NaN     NaN                NaN               NaN               NaN   
4     NaN     NaN                NaN               NaN               NaN   

  dia_chi_su_dung bluetooth  
0             NaN       NaN  
1             NaN       NaN  
2             NaN       NaN  
3             NaN       NaN  
4             NaN       NaN  

[5 rows x 84 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96647 entries, 0 to 96646
Data columns (total 84 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Unnamed: 0                    96647 non-null  int64  
 1   id                            96647 non-null  int64  
 2   master_id                     96647 non-null  int64  
 3   sku                           96647 non-null  int64  
 4   name                          96647 non-null  object 
 5   short_url                     96647 non-null  object 
 6   book_cover                    66029 non-null  object 
 7   price                         96647 non-null  int64  
 8   original_price                96647 non-null  int64  
 9   discount_rate                 96647 non-null  int64  
 10  rating_average                96647 non-null  float64
 11  review_count                  96647 non-null  int64  
 12  inventory_type                96647 non-null  object 
 13  p

***

## **Cleaning data**

Drop cột 'Unnamed: 0'

In [5]:
df = df.drop(columns='Unnamed: 0')
df.head()

id  master_id            sku  \
0  204317934  204317934  9972713357587   
1  205279963  205279963  7618382614299   
2  193209826  193209826  5001141446127   
3  205288789  205288789  9035774965791   
4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  rating_average  ...  kich_thuoc tai_trong  \
0          325000             19             5.0  ...         NaN       NaN   
1          325000             21             0.0  ...         NaN       NaN   
2          129000             26             4.8  ...         NaN       NaN   
3          325000             10             0.0  ...         NaN       NaN   
4          325000             10             0.0  ...         NaN       NaN   

  device_brand  do_nhay tan_so  huong_dan_bao_quan dac_diem_noi_bat  \
0          NaN      NaN    NaN                 NaN              NaN   
1          NaN      NaN    NaN                 NaN              NaN   
2          NaN      NaN    NaN                 NaN              NaN   
3          NaN      NaN    NaN                 NaN              NaN   
4          NaN      NaN    NaN                 NaN              NaN   

  dieu_kien_su_dung dia_chi_su_dung bluetooth  
0               NaN             NaN       NaN  
1               NaN             NaN       NaN  
2               NaN             NaN       NaN  
3               NaN             NaN       NaN  
4               NaN             NaN       NaN  

[5 rows x 83 columns]

Xem thử dữ liệu có bị trùng hay không?

In [6]:
df.duplicated().sum()

8294

Vậy là dữ liệu có 8294 dòng trùng nhau, ta sẽ tiến hành xoá những dòng này

In [7]:
df = df.drop_duplicates(ignore_index=True)
df.shape

(88353, 83)

#### **Mỗi dòng có ý nghĩa gì? Có các dòng có ý nghĩa khác nhau hay không?**

Mỗi dòng cho biết thông tin về một sản phẩm sách.

Vì khi thu thập dữ liệu, có những sản phẩm khác loại bị tiki xếp vào chung danh mục nên ta thấy những columns lạ bị thêm vào, ta phải tiến hành xoá những columns này và những sản phẩm (dòng) <font color =red>lạc loài </font>

In [8]:
df.columns

Index(['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'brand', 'brand_country', 'item_model_number',
       'origin', 'product_weight', 'battery_capacity', 'camera', 'chong_nuoc',
       'loai_day', 'san_pham', 'storage', 'battery_life', 'charge_time',
       'do_chiu_nuoc', 'included_accessories', 'material', 'luu_y',
       'minimum_inbound_policy_days', 'expiry_time', 'shelf_life_days',
       'chat_lieu', 'audio_power_output', 'audio_technology',
       'dieu_khien_bang_giong_noi', 'dieu_khien_tivi', 'digital_tivi_dvb_t2',
       'display_type', 'frequency', 'hdh_giao_dien',
       'image_processing_technology', 'network_hdmi', 'network_inte

In [9]:
df[~df['material'].isna()]

id  master_id            sku  \
13233  144356210  144356210  5525114755464   
14886  207035527  207035527  6816910771280   
15160  176689612  176689612  9588934896842   
25077  207631100  207631100  6801021851103   
49830  196565537  196565537  5928507424911   
63694  202432845  202432845  8183965466103   
76899  175150080  175150080  9992881446659   

                                                    name  \
13233  Túi đeo chạy bộ, túi đeo hông, đeo bụng dành c...   
14886  Áo Hoodie Nỉ Ngoại Nữ Thêu Chữ Loang Màu Hồng ...   
15160  Vỏ gối trắng trơn-vỏ gối cao cấp 100% Cotton-t...   
25077     Áo Khoác Nam Nữ Xanh Rêu Áo Khoác Dù Dáng Rộng   
49830  Rèm Tự Dính-Không Cần Khoan Vô Cùng Tiện Lợi, ...   
63694  BBSTORE'S Đầm Xòe Dự Tiệc Tay Phồng Thời Trang...   
76899  Bộ Bàn Ghế Gỗ Me Tây Nguyên Khối, Bàn Ghế Ăn G...   

                                               short_url book_cover     price  \
13233  https://tiki.vn/product-p144356210.html?spid=1...        NaN    112000   
14886  https://tiki.vn/product-p207035527.html?spid=2...        NaN    109000   
15160  https://tiki.vn/product-p176689612.html?spid=1...        NaN     63750   
25077  https://tiki.vn/product-p207631100.html?spid=2...        NaN    130000   
49830  https://tiki.vn/product-p196565537.html?spid=1...        NaN     52500   
63694  https://tiki.vn/product-p202432845.html?spid=2...        NaN    219000   
76899  https://tiki.vn/product-p175150080.html?spid=1...        NaN  20400000   

       original_price  discount_rate  rating_average  ...  \
13233          139000             19             4.8  ...   
14886          109000              0             0.0  ...   
15160           63750              0             0.0  ...   
25077          130000              0             0.0  ...   
49830          105000             50             0.0  ...   
63694          219000              0             0.0  ...   
76899        20400000              0             0.0  ...   

                 kich_thuoc tai_trong device_brand  do_nhay tan_so  \
13233                   NaN       NaN          NaN      NaN    NaN   
14886                   NaN       NaN          NaN      NaN    NaN   
15160                   NaN       NaN          NaN      NaN    NaN   
25077                   NaN       NaN          NaN      NaN    NaN   
49830           30 x 20 x 5       5kg          NaN      NaN    NaN   
63694                   NaN       NaN          NaN      NaN    NaN   
76899  3,39 x (70 - 75) x 5       NaN          NaN      NaN    NaN   

       huong_dan_bao_quan dac_diem_noi_bat dieu_kien_su_dung dia_chi_su_dung  \
13233                 NaN              NaN               NaN             NaN   
14886                 NaN              NaN               NaN             NaN   
15160                 NaN              NaN               NaN             NaN   
25077                 NaN              NaN               NaN             NaN   
49830                 NaN              NaN               NaN             NaN   
63694                 NaN              NaN               NaN             NaN   
76899                 NaN              NaN               NaN             NaN   

      bluetooth  
13233       NaN  
14886       NaN  
15160       NaN  
25077       NaN  
49830       NaN  
63694       NaN  
76899       NaN  

[7 rows x 83 columns]

Các `features` được sử dụng

In [10]:
features = ['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'luu_y']

Một số `features` không hợp lệ với yêu cầu bài toán

In [11]:
unvalid_cols = list(set(df.columns) - set(features))
print(unvalid_cols)

['tivi_type', 'brand', 'weight_without_stand', 'screen_mirroring', 'display_type', 'huong_dan_su_dung', 'frequency', 'battery_capacity', 'dac_diem_noi_bat', 'charge_time', 'do_chiu_nuoc', 'loai_pin', 'hdh_giao_dien', 'shelf_life_days', 'item_model_number', 'loai_day', 'kich_thuoc', 'battery_life', 'usb', 'san_pham', 'resolution', 'thoi_gian_su_dung', 'huong_dan_bao_quan', 'camera', 'size_without_stand_table_top', 'product_weight', 'digital_tivi_dvb_t2', 'image_processing_technology', 'chat_lieu', 'size_with_stand_table_top', 'dieu_khien_tivi', 'brand_country', 'remote_thong_minh', 'minimum_inbound_policy_days', 'network_hdmi', 'dia_chi_su_dung', 'storage', 'chong_nuoc', 'weight_with_stand', 'audio_power_output', 'dieu_khien_bang_giong_noi', 'network_wifi', 'origin', 'audio_technology', 'dieu_kien_su_dung', 'network_internet', 'do_nhay', 'expiry_time', 'tai_trong', 'material', 'size', 'included_accessories', 'device_brand', 'ung_dung', 'capacity', 'screen_size', 'bluetooth', 'tan_so']


Thử lấy các sản phẩm thuộc ngoài những `categories` của danh mục sách (mà ta đã thu thập)

In [12]:
valid_categories = pd.read_csv('id_data/categories_id.csv')['categories'] 
valid_categories = '|'.join(list(valid_categories)) #+ ['Root', 'Sách tiếng Việt', 'Nhà Sách Tiki'])
# df[df[unvalid_cols].notna().any(axis=1)].to_csv('not_books.csv', encoding='utf-8-sig')
# df[df['publisher_vn'].isna()].to_csv('not_publiser.csv', encoding='utf-8-sig')
# df[df['categories'].str.contains(valid_categories, na=False)]

#Chọn những categories không phải sách
df[~df['categories'].str.contains(valid_categories, na=False)]['categories'].value_counts()

Root                                   20440
Sách tiếng Việt                        19201
Nhà Sách Tiki                             23
Máy Ảnh - Máy Quay Phim                    4
Túi và Hộp Đựng Máy Ảnh                    3
Thiết Bị Số - Phụ Kiện Số                  3
Tai Nghe True Wireless                     2
Đầu Ghi Hình Camera                        1
Nước Ngọt                                  1
Móng giả                                   1
Bộ sản phẩm vệ sinh nhà cửa                1
Làm Đẹp - Sức Khỏe                         1
Thực phẩm khô khác                         1
Loa Vi Tính                                1
Dây Đeo Máy Ảnh, Máy Quay, Ống Nhòm        1
Bình, ly uống trà và phụ kiện              1
Voucher - Dịch vụ                          1
Bàn ghế làm việc                           1
Nón bơi                                    1
Phụ Kiện Máy Quay Khác                     1
Bao Da - Ốp Lưng iPad                      1
Bình đun siêu tốc                          1
Các loại g

Và ta chỉ lấy những giá trị hợp lệ 

In [13]:
df = df[df['categories'].str.contains(valid_categories, na=False)]
# df.shape
df.shape

(48291, 83)

Có thể vẫn còn, ta sẽ xem những features bị <font color=red>lạc loài</font> (khác giá trị nan):

In [14]:
unvalid_products = df[df[unvalid_cols].notna().any(axis=1)]
# display(unvalid_products)

Tiếp tục bỏ đi những sản phẩm này.

In [15]:
df = df.drop(unvalid_products.index).reset_index(drop=True)
df[df[unvalid_cols].notna().any(axis=1)] # test lại

Empty DataFrame
Columns: [id, master_id, sku, name, short_url, book_cover, price, original_price, discount_rate, rating_average, review_count, inventory_type, productset_group_name, day_ago_created, categories, all_time_quantity_sold, authors, publisher_vn, publication_date, dimensions, manufacturer, number_of_page, dich_gia, edition, brand, brand_country, item_model_number, origin, product_weight, battery_capacity, camera, chong_nuoc, loai_day, san_pham, storage, battery_life, charge_time, do_chiu_nuoc, included_accessories, material, luu_y, minimum_inbound_policy_days, expiry_time, shelf_life_days, chat_lieu, audio_power_output, audio_technology, dieu_khien_bang_giong_noi, dieu_khien_tivi, digital_tivi_dvb_t2, display_type, frequency, hdh_giao_dien, image_processing_technology, network_hdmi, network_internet, network_wifi, remote_thong_minh, resolution, screen_mirroring, screen_size, size_without_stand_table_top, size_with_stand_table_top, tivi_type, ung_dung, usb, weight_without_stand, weight_with_stand, size, loai_pin, thoi_gian_su_dung, capacity, huong_dan_su_dung, kich_thuoc, tai_trong, device_brand, do_nhay, tan_so, huong_dan_bao_quan, dac_diem_noi_bat, dieu_kien_su_dung, dia_chi_su_dung, bluetooth]
Index: []

[0 rows x 83 columns]

In [16]:
df.index.max()

48290

Có thể vẫn còn những sản phẩm <font color=red>lạc loài</font> khác, tuy nhiên khó có thể nhận biết và loại bỏ được.

Xóa đi các cột <font color =red> lạc loài </font> đi thôi!

In [17]:
df = df.drop(unvalid_cols, axis = 1)
df.shape

(48291, 25)

##### Kiểm tra lại thông tin của các cột trước khi bước qua **KHÁM PHÁ DỮ LIỆU**

> Mô tả thông tin của các cột


In [18]:
pd.read_csv("column/Columns.csv")

Name    Type  \
0                       id  string   
1                      sku  string   
2                     name  string   
3                short_url  string   
4               book_cover  string   
5                    price    int    
6           original_price     int   
7            discount_rate    int    
8           rating_average   float   
9             review_count     int   
10          inventory_type  string   
11   productset_group_name  string   
12         day_ago_created     int   
13              categories  string   
14  all_time_quantity_sold     int   
15                 authors  string   
16            publisher_vn  string   
17        publication_date    Date   
18              dimensions  string   
19            manufacturer  string   
20          number_of_page     int   
21                dich_gia  string   
22                 edition  string   
23                  luu_y   string   

                                             Describe  
0                                   ID của quyền sách  
1                                         Mã hàng hóa  
2                                      Tên quyển sách  
3                       Đường link đến trang bán sách  
4                                       Loại bìa sách  
5              Giá quyển sách đang được bán trên tiki  
6                              Giá gốc của quyển sách  
7                                         % Giảm giá   
8             Mức đánh giát trung bình ( 0 <= x <= 5)  
9          Số lượt đánh giá, bình luận của khách hàng  
10  Tình trạng trong kho của quyển sách. ("instock...  
11                                  Tên nhóm sản phẩm  
12               Số ngày kể từ khi đăng bán trên tiki  
13                                     Phân loại sách  
14                      Số lượng sách đã bán từ tiki   
15                                 Tác giả quyển sách  
16                                  Công ty phát hành  
17                                     Ngày xuất bản   
18        Kích thước chiều dài và rộng của quyển sách  
19                                       Nhà xuất bản  
20                                           Số trang  
21                                       Tên dịch giả  
22                           Phiên bản của quyển sách  
23                     Những lưu ý thay đổi trên sách

Cột `id` và `master_id` hoàn toàn giống nhau ở từng dòng. Vì thế ta sẽ xóa đi cột `master_id` <br>
Cột `publication_date` cần thay đổi `dtype` thành `datetime`

- Loại bỏ cột `master_id`

In [19]:
df = df.drop("master_id", axis = 1)

- Xem thông tin kiểu dữ liệu của mỗi cột

In [20]:
df.dtypes

id                          int64
sku                         int64
name                       object
short_url                  object
book_cover                 object
price                       int64
original_price              int64
discount_rate               int64
rating_average            float64
review_count                int64
inventory_type             object
productset_group_name      object
day_ago_created             int64
categories                 object
all_time_quantity_sold      int64
authors                    object
publisher_vn               object
publication_date           object
dimensions                 object
manufacturer               object
number_of_page             object
dich_gia                   object
edition                    object
luu_y                      object
dtype: object

Ta thấy cột `number_of_page` có kiểu dữ liệu là object chưa phù hợp (lẽ ra phải là interger), ta sẽ chuyển về đúng kiểu dữ liệu

In [21]:
# df['number_of_page'].fillna(0).astype(np.int64)
# df['number_of_page'].unique()
set(df['number_of_page'].apply(lambda x: type(x)))

{float, str}

Cột này chứa 2 kiểu dữ liệu là `str` (?) và `float`:
- Tại sao lại chứa kiểu dữ liệu string?

In [22]:
df[df['number_of_page'].apply(lambda x: isinstance(x, str))]['number_of_page'].unique()

array(['336', '178', '216', ..., '10512', '1058', '879'], dtype=object)

In [23]:
df[df['number_of_page'].str.contains(r'[a-zA-Z]', na=False)]['number_of_page'].unique()

array(['Cuốn', '80x2', 'mềm'], dtype=object)

- Thì ra các giá trị `integer` đã bị chuyển thành string.
- Một vài giá trị string <font color=red>lạc loài</font> ("Cuốn", "80x2", "mềm") đã làm kiểu dữ liệu bị thay đổi. Ta sẽ chuyển các giá trị này về nan.

In [24]:
df.loc[df['number_of_page'].str.contains(r'[a-zA-Z]', na=False), 'number_of_page'] = np.nan

In [25]:
df['number_of_page'] = df['number_of_page'].fillna(0).astype(np.int64)
df['number_of_page'].dtype

dtype('int64')

Đổi dtype của cột `publication_date` thành datetime

In [26]:
df.publication_date = pd.to_datetime(df.publication_date, format = "%Y-%m-%d %H:%M:%S", errors = "coerce")

Gần hơn 1 nửa dữ liệu bị lỗi hoặc thiếu datetime.

**Save dữ liệu đã xử lý để phục vụ cho việc xây dựng mô hình**

In [27]:
import os
os.mkdir('./features')
df.to_csv('./features/processed.csv',index=False, encoding='utf-8-sig')

FileExistsError: [WinError 183] Cannot create a file when that file already exists: './features'